ddl/ddl_api.go

```diff
@@ -3383,200 +3383,143 @@ func checkModifyCharsetAndCollation(toCharset, toCollate, origCharset, origColla
-			}
-			return unsupportedMsg, errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-		}
-		// Floating-point and fixed-point types also can be UNSIGNED. As with integer types, this attribute prevents
-		// negative values from being stored in the column. Unlike the integer types, the upper range of column values
-		// remains the same.
-		if to.Flen != origin.Flen || to.Decimal != origin.Decimal || mysql.HasUnsignedFlag(to.Flag) != mysql.HasUnsignedFlag(origin.Flag) {
-			msg := fmt.Sprintf("decimal change from decimal(%d, %d) to decimal(%d, %d)", origin.Flen, origin.Decimal, to.Flen, to.Decimal)
-			return msg, errUnsupportedModifyColumn.GenWithStackByArgs(msg)
-		}
-	case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp, mysql.TypeDuration, mysql.TypeYear:
-		switch origin.Tp {
-		case mysql.TypeDuration:
-			switch to.Tp {
-			case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
-				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-			}
-		case mysql.TypeDate:
-			switch to.Tp {
-			case mysql.TypeDatetime, mysql.TypeTimestamp:
-				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-			}
-		case mysql.TypeTimestamp:
-			switch to.Tp {
-			case mysql.TypeDuration, mysql.TypeDatetime:
-				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-			}
-		case mysql.TypeDatetime:
-			switch to.Tp {
-			case mysql.TypeTimestamp:
-				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-			}
-		case mysql.TypeYear:
-			switch to.Tp {
-			case mysql.TypeDuration:
-				return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-			case mysql.TypeYear:
-			default:
-				return "", ErrTruncatedWrongValue.GenWithStack("banned conversion that must fail")
+
+		if origin.Tp == mysql.TypeNewDecimal {
+			// Floating-point and fixed-point types also can be UNSIGNED. As with integer types, this attribute prevents
+			// negative values from being stored in the column. Unlike the integer types, the upper range of column values
+			// remains the same.
+			if to.Flen != origin.Flen || to.Decimal != origin.Decimal || mysql.HasUnsignedFlag(to.Flag) != mysql.HasUnsignedFlag(origin.Flag) {
+				msg := fmt.Sprintf("decimal change from decimal(%d, %d) to decimal(%d, %d)", origin.Flen, origin.Decimal, to.Flen, to.Decimal)
+				return true, msg, errUnsupportedModifyColumn.GenWithStackByArgs(msg)
 			}
 		}
-		switch to.Tp {
-		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp, mysql.TypeDuration:
-			skipSignCheck = true
-			skipLenCheck = true
-		case mysql.TypeYear:
-			if origin.Tp != mysql.TypeYear {
-				return "", ErrWarnDataOutOfRange.GenWithStack("banned conversion that must fail")
-			}
-			skipSignCheck = true
-			skipLenCheck = true
-		default:
-			return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
+
+		needReorg, reason := needReorgToChange(origin, to)
+		if !needReorg {
+			return
```

Comment:
> false "canReorg" means the types change are not supported.

Does this function return `false` but `nil` error is correct?

Reply:
This function return contains three scenarios:
1. false + nil err: types are compatible so can be directly changed
2. true + err: types are incompatible, and need reorg to change
3. false + err: types are incompatible, but change not supported

Actually the second scenario will be more clear if we return "true + nil err", but because CheckModifyTypeCompatible() is public and also called by `schema_amender.go:172`, the colChangeAmendable() logic seems depends on the error if type incompatible(no matter can reorg to change or not support).

Do you have any ideas about this?

Reply:
it ok, you can make comments on this function return value more clear like:

return nil err if types can be directly changed, or use `canReorg` indicate whether types can be changed by reorganization.

Reply:
Ok, that's good, thx.